EIPS/eip_generic_token_upgrade_standard.md

```diff
@@ -0,0 +1,353 @@
+``` solidity
+/// @dev A method to mock the upgrade call determining the amount of destination tokens received from an upgrade
+/// as well as the amount of source tokens that are left over as remainder
+/// @param sourceAmount The amount of source tokens that will be upgraded
+/// @return destinationAmount A uint256 representing the amount of destination tokens received if upgrade is called
+/// @return sourceRemainder A uint256 representing the amount of source tokens left over as remainder if upgrade is called
+function computeUpgrade(uint256 sourceAmount) external view
+        returns (uint256 destinationAmount, uint256 sourceRemainder)
+```
+
+#### computeDowngrade (Optional Ext.)
+
+Computes the `sourceAmount` of source tokens that correspond to a given `destinationAmount` of destination tokens, according to the predefined conversion ratio, as well as the `destinationRemainder` amount of destination tokens that can't be downgraded. For example, let's consider a (3, 2) ratio, which means that 3 destination tokens are provided for every 2 source tokens; for a destination amount of 13 tokens, `computeDowngrade(13)` should return `(4, 1)`, meaning that 4 source tokens are expected (in this case, from 12 destination tokens) and 1 destination token is left as remainder.
+``` solidity
+/// @dev A method to mock the downgrade call determining the amount of source tokens received from a downgrade
+/// as well as the amount of destination tokens that are left over as remainder
+/// @param destinationAmount The amount of destination tokens that will be downgraded
+/// @return sourceAmount A uint256 representing the amount of source tokens received if downgrade is called
+/// @return destinationRemainder A uint256 representing the amount of destination tokens left over as remainder if upgrade is called
+function computeDowngrade(uint256 destinationAmount) external view
+        returns (uint256 sourceAmount, uint256 destinationRemainder)
+```
+
+
+#### upgrade
+
+Upgrades the `amount` of source token to the destination token in the specified ratio. The destination tokens will be sent to the ```_to``` address. The function should lock the source tokens in the upgrade contract or burn them. If the downgrade Optional Ext. is implemented, the source tokens should be locked instead of burning. The function should `throw` if the caller's address does not have enough source token to upgrade or if ```isUpgradeActive``` is returning ```false```. The function MUST also fire the `Upgrade` event. `approve` Must be called first on the source contract. 
```

Comment:
> If the downgrade Optional Ext. is implemented, the source tokens should be locked instead of burning.

I think this is inappropriate behavior as it creates an asymmetry.  Ideally, the sum of `totalSupply` on both tokens will equal the total supply of the overall upgradable token.  This holds for burnable tokens that are upgradable but not downgradable, but it does not hold for burnable tokens which are upgradable and downgradable.

I think the correct behavior here is to always burn when possible (side note: burning should be defined here as adjusting the total supply down).

A mechanism for identifying whether a given token is burnable or not I think should also be included as it allows the person looking at the contract to determine whether or not they can rely on `totalSupply` summing up to something reasonable or not, and the specification should assert that if a token is burnable, tokens **MUST** be burned (rather than **SHOULD** be burned) and it should also strongly recommend that tokens be burnable whenever possible (**SHOULD**).

On a similar note, it should be more strict about transferring to `0` address and assert that if the token is not burnable, then it **MUST** transfer to `0`.  This way people interacting with the token can calculate the total supply across upgrades reliably, regardless of whether it is burnable or not.

Reply:
I suppose one issue with the above is that if a source token already exists and isn't mintable, and you want downgrades, then you need the tokens to remain accessible by the upgrader contract.  I think this convinces me that the current recommendations are acceptable, but I think they should be made strict with **MUST** rather than **SHOULD**.

I think it may actually be best to just assert that the source token is *always* locked in the upgrade contract, so the behavior is consistent.  The problem of course is that `totalSupply` will be wrong in that case which is unfortunate.  Maybe there can be a method exposed that indicates whether total supply is `source.totalSupply + destination.totalSupply` or `source.totalSupply + destination.totalSupply - balenceOf(upgrader)`?  Or perhaps the upgrader should just be required to expose a `totalSupply` function that can have whatever logic it wants in it?

Reply:
I agree that the language surrounding this should be more strict (SHOULD -> MUST)! I will go ahead and commit this change. As to the proposed assertion that the source token is always to be locked in the upgrade contract, I agree that it would make behavior more consistent but opens up several issues with strictness as well as total supply as you mentioned. We will need to discuss the implications of this further before making that change.